### PR TITLE
Visitor for stable rust

### PR DESCRIPTION
**Pull request by kdy1**
I've implemented a Visitor for stable rustc.
The implementation is naive, but it works. 

Generated code looks like

```rust

pub trait Visit {
   #[allow(unused_variables)]
    fn visit_array_lit(&self, n: &ArrayLit, _parent: &dyn Node) {
        match n {
            ArrayLit { span, elems } => {
                self.visit_span(&*span, n as _);
                self.visit_opt_vec_expr_or_spreads(&*elems, n as _);
            }
        }
    }
    // .. LOTS of generated methods
    #[allow(unused_variables)]
    fn visit_decl(&self, n: &Decl, _parent: &dyn Node) {
        match n {
            Decl::Class { 0: _0 } => {
                self.visit_class_decl(&*_0, n as _);
            }
            Decl::Fn { 0: _0 } => {
                self.visit_fn_decl(&*_0, n as _);
            }
            Decl::Var { 0: _0 } => {
                self.visit_var_decl(&*_0, n as _);
            }
            Decl::TsInterface { 0: _0 } => {
                self.visit_ts_interface_decl(&*_0, n as _);
            }
            Decl::TsTypeAlias { 0: _0 } => {
                self.visit_ts_type_alias_decl(&*_0, n as _);
            }
            Decl::TsEnum { 0: _0 } => {
                self.visit_ts_enum_decl(&*_0, n as _);
            }
            Decl::TsModule { 0: _0 } => {
                self.visit_ts_module_decl(&*_0, n as _);
            }
        }
    }
     // .. LOTS of generated methods
}
```

where `Node` is declared as

```rust
pub trait Node: Any {}
impl<T: ?Sized> Node for T where T: Any {}
```

cc @ry @bartlomieju

### CODE REVIEW

**Comment by bartlomieju** [comment]
@kdy1 I tried `Visit` in local checkout of your branch, a few questions arise:
Say I want to override some method to do custom processing:
```rust
impl Visit for LintContext {
  fn visit_module(&self, module: &Module, parent: &dyn Node) {
    assert!(self.scope_stack.is_empty());

    let module_node = Box::new(AstNode::Module(module.clone()));
    let current_scope = self.get_current_scope();

    let module_scope = Rc::new(RefCell::new(Scope::new(
      module_node.clone(),
      ScopeKind::Module,
      Some(current_scope.clone()),
    )));

    // Entering module scope
    self.scope_stack.push(module_scope);

    // Call default implementation and walk children
    // NOTE: this doesn't work 
    Visit::visit_module(self, module, parent);

    // Exiting module scope
    self.scope_stack.pop();
  }
}
```
How do I go about calling default implementation of `visit_module`(AFAIK that's not possible in Rust)?
Is it possible to do something like `self.visit_children(module)`?

**Comment by kdy1** [comment]
@bartlomieju It can be solved by extracting default behavior as a function.

It would be like

```rust
pub trait Visit {
    fn visit_module(&mut self, n: &Module, parent: &dyn Node) {
        visit_module(self, n, parent)
    }
}

pub fn visit_module<V>(v: &mut V, module: &Module, parent: &Node) {
     // default implementation lives here
}
```

I'll do it tomorrow (well, actually today for me as it's over 12pm).

**Comment by bartlomieju** [comment]
@kdy1 awesome, thank you!

**Comment by kdy1** [comment]
@bartlomieju I did it. Can you check it, please?

File: ecmascript/visit/macros/src/lib.rs [hunk 1]
```diff
@@ -74,6 +75,46 @@ pub fn define(tts: proc_macro::TokenStream) -> proc_macro::TokenStream {
...
+
+        tokens.push_tokens(&q!(
+            Vars {
+                fn_name,
+                default_body,
+                Type: arg_ty,
+            },
+            {
+                #[allow(unused_variables)]
+                fn fn_name<V: ?Sized + Visit>(_visitor: &mut V, n: Type, _parent: &dyn Node) {
```
Comment (bartlomieju) [inline ecmascript/visit/macros/src/lib.rs:113]:
```suggestion
                pub fn fn_name<V: ?Sized + Visit>(_visitor: &mut V, n: Type, _parent: &dyn Node) {
```
With this change it's working as expected 👍

Comment (kdy1) [reply ecmascript/visit/macros/src/lib.rs]: Thanks for the review. I forgot exporting it.

**Comment by bartlomieju** [comment]
@kdy1 any chance you could release it today?

**Comment by kdy1** [comment]
@bartlomieju I published it.

**Comment by bartlomieju** [comment]
@kdy1 kudos!

**Pull request merged**